File: test/generic/generic-01.cpp
/// @file

/// @brief
/// @param foo (T) generic T
/// @param x (integer) 
/// @returns (T) generic T
template <typename T>
(T) gen01(T foo, integer x);

/// @brief
/// @param foo (T❲❳) generic array
/// @returns (T❲❳) generic array
template <typename T>
(T❲❳) gen02(T❲❳ foo);

/// @brief
/// @param foo (T→number) 
/// @returns (T→number) 
template <typename T→number>
(T→number) gen03(T→number foo);

/// @brief
/// @param foo (T1) 
/// @param bar (T2) 
/// @param xox (T3) 
template <typename T1, typename T2, typename T3>
gen04(T1 foo, T2 bar, T3 xox);

File: test/generic/generic-01.lua
---@diagnostic disable:lowercase-global
---@file

---@generic T
---@param foo T # generic T
---@param x integer
---@return T # generic T
function gen01(foo, x)
   return foo
end


---@generic T
---@param foo T[] # generic array
---@return T[] # generic array
function gen02(foo)
   return foo
end


---@generic T : number
---@param foo T
---@return T
function gen03(foo)
   return foo
end


---@generic T1
---@generic T2
---@generic T3
---@param foo T1
---@param bar T2
---@param xox T3
function gen04(foo, bar, xox)
end
